crates/commitlog/src/commit.rs

```diff
@@ -0,0 +1,234 @@
+
+    /// The largest transaction offset in this commit.
+    pub fn max_tx_offset(&self) -> u64 {
+        self.min_tx_offset + self.n as u64
+    }
+
+    /// Length in bytes of this commit when written to the log via [`Self::write`].
+    pub fn encoded_len(&self) -> usize {
+        Self::FRAMING_LEN + self.records.len()
+    }
+
+    /// Serialize and write `self` to `out`.
+    pub fn write<W: Write>(&self, out: W) -> io::Result<()> {
+        let mut out = Crc32cWriter::new(out);
+
+        let min_tx_offset = self.min_tx_offset.to_le_bytes();
+        let n = self.n.to_le_bytes();
+        let len = (self.records.len() as u32).to_le_bytes();
+
+        out.write_all(&min_tx_offset)?;
+        out.write_all(&n)?;
+        out.write_all(&len)?;
+        out.write_all(&self.records)?;
+
+        let crc = out.crc32c();
+        let mut out = out.into_inner();
+        out.write_all(&crc.to_le_bytes())?;
+
+        Ok(())
+    }
+
+    /// Attempt to read one [`Commit`] from the given [`Read`]er.
+    ///
+    /// Returns `None` if the reader is already at EOF.
+    ///
+    /// Verifies the checksum of the commit. If it doesn't match, an error of
+    /// kind [`io::ErrorKind::InvalidData`] with an inner error downcastable to
+    /// [`ChecksumMismatch`] is returned.
+    pub fn decode<R: Read>(reader: R) -> io::Result<Option<Self>> {
+        let mut reader = Crc32cReader::new(reader);
+
+        let Some(hdr) = Header::decode(&mut reader)? else {
+            return Ok(None);
+        };
+        let mut records = vec![0; hdr.len as usize];
+        reader.read_exact(&mut records)?;
+
+        let chk = reader.crc32c();
+        let crc = decode_u32(reader.into_inner())?;
+
+        if chk != crc {
+            return Err(invalid_data(ChecksumMismatch));
+        }
+
+        Ok(Some(Self {
+            min_tx_offset: hdr.min_tx_offset,
+            n: hdr.n,
+            records,
+        }))
+    }
+}
+
+/// Numbers needed to compute [`crate::segment::Header`].
+#[derive(Clone, Copy, Debug, Eq, PartialEq)]
+pub struct Metadata {
+    pub min_tx_offset: u64,
+    pub max_tx_offset: u64,
+    pub size_in_bytes: u64,
+}
+
+impl Metadata {
+    /// Extract the [`Metadata`] of a single [`Commit`] from the given reader.
+    ///
+    /// Note that this decodes the commit due to checksum verification.
+    /// Like [`Commit::decode`], returns `None` if the reader is at EOF already.
+    pub fn extract<R: io::Read>(reader: R) -> io::Result<Option<Self>> {
+        Commit::decode(reader).map(|maybe_commit| maybe_commit.map(Self::from))
+    }
+}
+
+impl From<Commit> for Metadata {
+    fn from(
+        Commit {
+            min_tx_offset,
+            n,
+            records,
+        }: Commit,
+    ) -> Self {
+        Self {
+            min_tx_offset,
+            max_tx_offset: min_tx_offset + n as u64,
+            size_in_bytes: Header::LEN as u64 + records.len() as u64 + /* crc32 */ 4,
```

Comment:
Shouldn't this just be a function on the `Commit`?

Reply:
#985